### PR TITLE
gh-91172: Create a workflow for verifying bundled pip and setuptools

### PR DESCRIPTION
**Pull request by illia-v**
https://bugs.python.org/issue47016

### CODE REVIEW

File: .github/workflows/ensurepip_verify_bundled.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      - 'Lib/ensurepip/_bundled/**'
+  pull_request:
+    paths:
+      - 'Lib/ensurepip/_bundled/**'
+
+jobs:
+  verify:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (hugovk) [inline .github/workflows/ensurepip_verify_bundled.yml]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/ensurepip_verify_bundled.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      - name: Compare checksums of bundled pip and setuptools to ones published on PyPI
+        run: |
+          package_names=("pip" "setuptools")
```
Comment (hugovk) [inline .github/workflows/ensurepip_verify_bundled.yml]: Perhaps this could be moved to a script (say, under `Tools/scripts`), so it can also be run locally?

Comment (illia-v) [reply .github/workflows/ensurepip_verify_bundled.yml]: Good idea, but it looks like all scripts in `Tools/scripts` are written in Python.
This is a Bash script, and it may not work on some platforms.

Do you know how compatible with different platforms scripts located in `Tools/scripts` should be?

Comment (hugovk) [reply .github/workflows/ensurepip_verify_bundled.yml]: Hmm, or perhaps the `Misc` directory? The [devguide](https://devguide.python.org/?highlight=misc#additional-resources) says:

> Various tools with configuration files as found in the [Misc directory](https://github.com/python/cpython/tree/main/Misc)

I don't know what compatibility is needed, but as this script is primarily intended for Ubuntu on the CI, I'd stick with that.

Comment (illia-v) [reply .github/workflows/ensurepip_verify_bundled.yml]: Done. [This is](https://github.com/illia-v/cpython/runs/5557015587?check_suite_focus=true) the latest run after I corrupted the pip wheel temporarily.

BTW, I added messages that GitHub has to show as annotations. But I could not see the annotations, maybe it does not show them for binary files…

Comment (hugovk) [reply .github/workflows/ensurepip_verify_bundled.yml]: They show up on the Summary page, for example clicking "🏠 Summary" at the top left: 

image

 👍

File: .github/workflows/verify_bundled_wheels.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+name: Verify bundled pip and setuptools
```
Comment (hugovk) [inline .github/workflows/verify_bundled_wheels.yml]: Minor suggestion: I know the only other multi-word yml filename in https://github.com/python/cpython/tree/main/.github/workflows uses an underscore:

```console
$ ls .github/workflows
build.yml         build_msi.yml     doc.yml           posix-deps-apt.sh stale.yml
```

But I'd probably go for dashes here, like `posix-deps-apt.sh`.

**Review by hugovk — APPROVED** [review]
Thank you! 🕊️🇺🇦✌️

File: .github/workflows/verify_bundled_wheels.yml (hunk 3 above)
Comment (illia-v) [reply .github/workflows/verify_bundled_wheels.yml]: Thank you for the review!

> But I'd probably go for dashes here, like `posix-deps-apt.sh`.

I renamed the file.

**Review by pradyunsg — APPROVED** [review]

File: Misc/verify-bundled-wheels.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
```
Comment (AA-Turner) [inline Misc/verify-bundled-wheels.sh]: Is there a reason this is written as a shell script? If (as @hugovk mentioned) it would be useful to run locally, rewriting in Python would make this more portable cross-platform (e.g. Windows).

Happy to help rewriting, just don't want to do the work if there's a requirement to have it be a shell script I'm missing!

A

Comment (illia-v) [reply Misc/verify-bundled-wheels.sh]: > Is there a reason this is written as a shell script?

[Initially](https://github.com/python/cpython/pull/31885/commits/e4a8faf7b4e4aeb553e246d7e58d887a018dc341#diff-d7623692dd445a8dd8e63aaa1a557c50c8cd4e530aee2634865caa0b402a69a8) this script was embedded in the workflow file, making it a shell script seemed to be cleaner at the time.

Comment (pradyunsg) [reply Misc/verify-bundled-wheels.sh]: I think this can be a Python script, but changing over to that should not be a blocking concern for the PR. :)

**Comment by AA-Turner** [comment]
@illia-v the below is my attempt at a Python rewrite:

```python
"""
Compare checksums of bundled pip and setuptools to ones
published on PyPI (retrieved via the Warehouse’s JSON API).
"""

import hashlib
import json
from pathlib import Path
import re
from urllib.request import urlopen

PACKAGE_NAMES = ("pip", "setuptools")
ROOT = Path(__file__).parent.parent / "Lib/ensurepip"
WHEEL_DIR = ROOT / "_bundled"
ENSURE_PIP_INIT_PY_TEXT = (ROOT / "__init__.py").read_text(encoding="utf-8")

# Contains names of successfully verified packages
verified_packages = {*()}

for package_name in PACKAGE_NAMES:
    # Find the package on disk
    package_path = next(WHEEL_DIR.glob(f"{package_name}*.whl"), None)
    if package_path is None:
        continue

    print(f"Verifying checksum for {package_path}.")

    # Find the version of the package used by ensurepip
    package_version_match = re.search(
        f'_{package_name.upper()}_VERSION = "([^"]+)',
        ENSURE_PIP_INIT_PY_TEXT
    )
    if package_version_match is None:
        continue
    package_version = package_version_match[1]

    # Get the SHA 256 digest from the Cheeseshop
    try:
        raw_text = urlopen(f"https://pypi.org/pypi/{package_name}/json").read()
    except (OSError, ValueError) as error:
        continue

    expected_digest = ""
    release_files = json.loads(raw_text)["releases"][package_version]
    for release_info in release_files:
        if package_path.name != release_info["filename"]:
            continue
        expected_digest = release_info["digests"].get("sha256", "")
    if expected_digest == "":
        continue

    # Compute the SHA 256 digest of the wheel on disk
    actual_digest = hashlib.sha256(package_path.read_bytes()).hexdigest()

    print(f"Expected digest: {expected_digest}")
    print(f"Actual digest:   {actual_digest}")

    # The messages are formatted to be parsed by GitHub Actions.
    # https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-notice-message
    if actual_digest == expected_digest:
        print(f"::notice file={package_path}::"
              f"Successfully verified the checksum of the {package_name} wheel.\n")
        verified_packages.add(package_name)
    else:
        print(f"::error file={package_path}::"
              f"Failed to verify the checksum of the {package_name} wheel.\n")

# If we verified all packages, the set of package names and the set of verified
# packages will be equal.
if {*PACKAGE_NAMES} == verified_packages:
    raise SystemExit(0)

# Otherwise we failed to verify all the packages.
raise SystemExit(1)
```

A

**Comment by AA-Turner** [comment]
@illia-v the below is my attempt at a Python rewrite:

```python
"""Compares checksums for wheels in :mod:`ensurepip` against the Cheeseshop."""

import hashlib
import json
from pathlib import Path
import re
from urllib.request import urlopen

PACKAGE_NAMES = ("pip", "setuptools")
ROOT = Path(__file__).parent.parent / "Lib/ensurepip"
WHEEL_DIR = ROOT / "_bundled"
ENSURE_PIP_INIT_PY_TEXT = (ROOT / "__init__.py").read_text(encoding="utf-8")

# Contains names of successfully verified packages
verified_packages = {*()}

for package_name in PACKAGE_NAMES:
    # Find the package on disk
    package_path = next(WHEEL_DIR.glob(f"{package_name}*.whl"), None)
    if package_path is None:
        continue

    print(f"Verifying checksum for {package_path}.")

    # Find the version of the package used by ensurepip
    package_version_match = re.search(
        f'_{package_name.upper()}_VERSION = "([^"]+)',
        ENSURE_PIP_INIT_PY_TEXT
    )
    if package_version_match is None:
        continue
    package_version = package_version_match[1]

    # Get the SHA 256 digest from the Cheeseshop
    try:
        raw_text = urlopen(f"https://pypi.org/pypi/{package_name}/json").read()
    except (OSError, ValueError) as error:
        continue

    expected_digest = ""
    release_files = json.loads(raw_text)["releases"][package_version]
    for release_info in release_files:
        if package_path.name != release_info["filename"]:
            continue
        expected_digest = release_info["digests"].get("sha256", "")
    if expected_digest == "":
        continue

    # Compute the SHA 256 digest of the wheel on disk
    actual_digest = hashlib.sha256(package_path.read_bytes()).hexdigest()

    print(f"Expected digest: {expected_digest}")
    print(f"Actual digest:   {actual_digest}")

    # The messages are formatted to be parsed by GitHub Actions.
    # https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-notice-message
    if actual_digest == expected_digest:
        print(f"::notice file={package_path}::"
              f"Successfully verified the checksum of the {package_name} wheel.\n")
        verified_packages.add(package_name)
    else:
        print(f"::error file={package_path}::"
              f"Failed to verify the checksum of the {package_name} wheel.\n")

# If we verified all packages, the set of package names and the set of verified
# packages will be equal.
if {*PACKAGE_NAMES} == verified_packages:
    raise SystemExit(0)

# Otherwise we failed to verify all the packages.
raise SystemExit(1)
```

A

**Comment by pradyunsg** [comment]
> BTW, isn't Warehouse the new name of what was called Cheese Shop?

Well, Python Package Index used to be called Cheese Shop (based on a Monty Python sketch about a Cheese Shop that didn't have any cheese) -- the current codebase for PyPI is called Warehouse: https://github.com/pypa/warehouse/

**Comment by AA-Turner** [comment]
> Could you please push it to this branch or create a pull request to it?

I don't have permission to push, I'll create a PR.

A

**Comment by AA-Turner** [comment]
https://github.com/illia-v/cpython/pull/1

A

**Comment by illia-v** [comment]
@AA-Turner thanks for converting the script to Python!
https://github.com/python/cpython/actions/runs/2455851364 is an example run with a corrupted wheel.

File: Tools/scripts/verify_ensurepip_wheels.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    print_notice(
+        package_path,
+        f"Successfully verified the checksum of the {package_name} wheel.",
+    )
+    return True
+
+
+if __name__ == "__main__":
+    success = all([verify_wheel(package_name) for package_name in PACKAGE_NAMES])
+    sys.exit(0 if success else 1)
```
Comment (AA-Turner) [inline Tools/scripts/verify_ensurepip_wheels.py]: Don't construct an inner list, eliminate the `sys` import:
```suggestion
    success = all(verify_wheel(package_name) for package_name in PACKAGE_NAMES)
    if success:
        raise SystemExit(0)
    raise SystemExit(1)
```

File: Tools/scripts/verify_ensurepip_wheels.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        break
+    else:
+        print_error(package_path, f"No digest for {package_name} found from PyPI.")
+        return False
+
+    # Compute the SHA 256 digest of the wheel on disk
+    actual_digest = hashlib.sha256(package_path.read_bytes()).hexdigest()
+
+    print(f"Expected digest: {expected_digest}")
+    print(f"Actual digest:   {actual_digest}")
```
Comment (AA-Turner) [inline Tools/scripts/verify_ensurepip_wheels.py:78]:
```suggestion
    print_notice(package_path, f"Expected digest: {expected_digest}")
    print_notice(package_path, f"Actual digest:   {actual_digest}")
```

File: Tools/scripts/verify_ensurepip_wheels.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def verify_wheel(package_name: str) -> bool:
+    # Find the package on disk
+    package_path = next(WHEEL_DIR.glob(f"{package_name}*.whl"), None)
+    if not package_path:
+        print_error("", f"Could not find a {package_name} wheel on disk.")
+        return False
+
+    print(f"Verifying checksum for {package_path}.")
```
Comment (AA-Turner) [inline Tools/scripts/verify_ensurepip_wheels.py:43]:
```suggestion
    print_notice(package_path, f"Verifying checksum for {package_path}.")
```

File: Tools/scripts/verify_ensurepip_wheels.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import re
+import sys
+from pathlib import Path
+from urllib.request import urlopen
+
+PACKAGE_NAMES = ("pip", "setuptools")
+ENSURE_PIP_ROOT = Path(__file__).parent.parent.parent / "Lib/ensurepip"
+WHEEL_DIR = ENSURE_PIP_ROOT / "_bundled"
+ENSURE_PIP_INIT_PY_TEXT = (ENSURE_PIP_ROOT / "__init__.py").read_text(encoding="utf-8")
+GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS")
```
Comment (AA-Turner) [inline Tools/scripts/verify_ensurepip_wheels.py]: For explicitness?
```suggestion
GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS", False)
```

File: Misc/NEWS.d/next/Tests/2022-03-14-23-28-17.bpo-47016.K-t2QX.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Create a GitHub Actions workflow for verifying bundled pip and setuptools.
+Patch by Illia Volochii and Adam Turner.
```
Comment (AA-Turner) [inline Misc/NEWS.d/next/Tests/2022-03-14-23-28-17.bpo-47016.K-t2QX.rst:2]: Thanks!

**Review by AA-Turner — COMMENTED** [review]
A few comments (I like the function refactoring!)

A

File: Tools/scripts/verify_ensurepip_wheels.py (hunk 5 above)
Comment (illia-v) [reply Tools/scripts/verify_ensurepip_wheels.py]: The list comprehension was needed to verify both wheels even if the first verification fails. I've made this piece of code more explicit.

File: Tools/scripts/verify_ensurepip_wheels.py (hunk 6 above)
Comment (illia-v) [reply Tools/scripts/verify_ensurepip_wheels.py:78]: If I make the change, this will add redundant annotation bloat.

I think it is enough to state that a wheel has either been verified or not like in this two cases:
* https://github.com/illia-v/cpython/actions/runs/2455864579
* https://github.com/python/cpython/actions/runs/2455851364

Comment (AA-Turner) [reply Tools/scripts/verify_ensurepip_wheels.py:78]: Ahh, I didn't realise it was that prominent -- in that case makes sense to keep as current.

A

File: Tools/scripts/verify_ensurepip_wheels.py (hunk 8 above)
Comment (illia-v) [reply Tools/scripts/verify_ensurepip_wheels.py]: I believe checking its presence is enough. On GitHub Actions the value will be `"true"` as a string. I don't think it is worth changing it to `GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"` either.

Comment (AA-Turner) [reply Tools/scripts/verify_ensurepip_wheels.py]: In that case, maybe `"GITHUB_ACTIONS" in os.environ` -- makes it clear that we don't care about the value at all.

A

Comment (illia-v) [reply Tools/scripts/verify_ensurepip_wheels.py]: On the second thought, I changed it to `GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"` to be explicit.

File: .github/workflows/verify-ensurepip-wheels.yml [hunk 10]
```diff
@@ -17,6 +17,5 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: actions/setup-python@v3
```
Comment (AA-Turner) [inline .github/workflows/verify-ensurepip-wheels.yml]: Why was this removed? We should bump to v4:

```yaml
- uses: actions/setup-python@v4
  with:
    python-version: '3'
```

File: Tools/scripts/verify_ensurepip_wheels.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import os
+import re
+from pathlib import Path
+from urllib.request import urlopen
+
+PACKAGE_NAMES = ("pip", "setuptools")
+ENSURE_PIP_ROOT = Path(__file__).parent.parent.parent / "Lib/ensurepip"
+WHEEL_DIR = ENSURE_PIP_ROOT / "_bundled"
+ENSURE_PIP_INIT_PY_TEXT = (ENSURE_PIP_ROOT / "__init__.py").read_text(encoding="utf-8")
+GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"
```
Comment (AA-Turner) [inline Tools/scripts/verify_ensurepip_wheels.py:21]: A simpler check:
```suggestion
GITHUB_ACTIONS = "GITHUB_ACTIONS" in os.environ
```

File: Tools/scripts/verify_ensurepip_wheels.py (hunk 8 above)
Comment (AA-Turner) [reply Tools/scripts/verify_ensurepip_wheels.py]: Sorry, missed this comment -- I still think the simpler check is better.

A

File: .github/workflows/verify-ensurepip-wheels.yml (hunk 10 above)
Comment (illia-v) [reply .github/workflows/verify-ensurepip-wheels.yml]: I could either make the change you suggest or use Python from ubuntu-latest that is enough, I chose the latter.

Comment (illia-v) [reply .github/workflows/verify-ensurepip-wheels.yml]: https://github.com/python/cpython/runs/6938767354?check_suite_focus=true verifies that this workflow works without `actions/setup-python`.

Comment (AA-Turner) [reply .github/workflows/verify-ensurepip-wheels.yml]: I agree that it works, but it loses consistency with all the other workflows that use (rather than build) python, and ties us to the Python update cadence that Ubuntu use. I would argue that it is also more discoverable on the GHA UI, rather than implicitly using Ubuntu's default Python.

A

Comment (pradyunsg) [reply .github/workflows/verify-ensurepip-wheels.yml]: Besides, unlike Ubuntu's Python, setup-python's builds don't have esoteric patches that deviate behaviour in various ways.

Comment (illia-v) [reply .github/workflows/verify-ensurepip-wheels.yml]: Fair enough

**Review by AA-Turner — APPROVED** [review]

**Comment by AA-Turner** [comment]
Although this is a new workflow file I've added the backport labels as I think we should be verifying the backport PRs for `ensurepip`.

A
